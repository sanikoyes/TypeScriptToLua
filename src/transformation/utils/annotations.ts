import * as ts from "typescript";
import { TransformationContext } from "../context";

export enum AnnotationKind {
    Extension = "extension",
    MetaExtension = "metaExtension",
    CustomConstructor = "customConstructor",
    CompileMembersOnly = "compileMembersOnly",
    NoResolution = "noResolution",
    PureAbstract = "pureAbstract",
    Phantom = "phantom",
    TupleReturn = "tupleReturn",
    LuaIterator = "luaIterator",
    LuaTable = "luaTable",
    NoSelf = "noSelf",
    NoSelfInFile = "noSelfInFile",
    Vararg = "vararg",
    ForRange = "forRange",
    Jsx = "jsx",
    JsxFrag = "jsxFrag",
}

export interface Annotation {
    kind: AnnotationKind;
    args: string[];
}

function createAnnotation(name: string, args: string[]): Annotation | undefined {
    const kind = Object.values(AnnotationKind).find(k => k.toLowerCase() === name.toLowerCase());
    if (kind !== undefined) {
        return { kind, args };
    }
}

export type AnnotationsMap = Map<AnnotationKind, Annotation>;

function collectAnnotations(source: ts.Symbol | ts.Signature, annotationsMap: AnnotationsMap): void {
    for (const tag of source.getJsDocTags()) {
        const annotation = createAnnotation(tag.name, tag.text?.map(p => p.text) ?? []);
        if (annotation) {
            annotationsMap.set(annotation.kind, annotation);
        }
    }
}

export function getSymbolAnnotations(symbol: ts.Symbol): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();
    collectAnnotations(symbol, annotationsMap);
    return annotationsMap;
}

export function getTypeAnnotations(type: ts.Type): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    if (type.symbol) collectAnnotations(type.symbol, annotationsMap);
    if (type.aliasSymbol) collectAnnotations(type.aliasSymbol, annotationsMap);

    return annotationsMap;
}

export function getNodeAnnotations(node: ts.Node): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    for (const tag of ts.getJSDocTags(node)) {
        const tagName = tag.tagName.text;
        const annotation = createAnnotation(tagName, getTagArgsFromComment(tag));
        if (annotation) {
            annotationsMap.set(annotation.kind, annotation);
        }
    }

    return annotationsMap;
}

export function getFileAnnotations(sourceFile: ts.SourceFile): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();

    if (sourceFile.statements.length > 0) {
        // Manually collect jsDoc because `getJSDocTags` includes tags only from closest comment
        const jsDoc = sourceFile.statements[0].jsDoc;
        if (jsDoc) {
            for (const tag of jsDoc.flatMap(x => x.tags ?? [])) {
                const tagName = tag.tagName.text;
                const annotation = createAnnotation(tagName, getTagArgsFromComment(tag));
                if (annotation) {
                    annotationsMap.set(annotation.kind, annotation);
                }
            }
        }
    }

    return annotationsMap;
}

export function getSignatureAnnotations(context: TransformationContext, signature: ts.Signature): AnnotationsMap {
    const annotationsMap: AnnotationsMap = new Map();
    collectAnnotations(signature, annotationsMap);

    // Function properties on interfaces have the JSDoc tags on the parent PropertySignature
    const declaration = signature.getDeclaration();
    if (declaration?.parent && ts.isPropertySignature(declaration.parent)) {
        const symbol = context.checker.getSymbolAtLocation(declaration.parent.name);
        if (symbol) {
            collectAnnotations(symbol, annotationsMap);
        }
    }

    return annotationsMap;
}

export function isTupleReturnCall(context: TransformationContext, node: ts.Node): boolean {
    if (!ts.isCallExpression(node)) {
        return false;
    }

    const signature = context.checker.getResolvedSignature(node);
    if (signature) {
        if (getSignatureAnnotations(context, signature).has(AnnotationKind.TupleReturn)) {
            return true;
        }

        // Only check function type for directive if it is declared as an interface or type alias
        const declaration = signature.getDeclaration();
        const isInterfaceOrAlias =
            declaration?.parent &&
            ((ts.isInterfaceDeclaration(declaration.parent) && ts.isCallSignatureDeclaration(declaration)) ||
                ts.isTypeAliasDeclaration(declaration.parent));
        if (!isInterfaceOrAlias) {
            return false;
        }
    }

    const type = context.checker.getTypeAtLocation(node.expression);
    return getTypeAnnotations(type).has(AnnotationKind.TupleReturn);
}

export function isLuaIteratorType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.LuaIterator);
}

export function isVarargType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.Vararg);
}

export function isForRangeType(context: TransformationContext, node: ts.Node): boolean {
    const type = context.checker.getTypeAtLocation(node);
    return getTypeAnnotations(type).has(AnnotationKind.ForRange);
}

export function getTagArgsFromComment(tag: ts.JSDocTag): string[] {
    if (tag.comment) {
        if (typeof tag.comment === "string") {
            return tag.comment.split(" ");
        } else {
            return tag.comment.map(part => part.text);
        }
    }

    return [];
}
