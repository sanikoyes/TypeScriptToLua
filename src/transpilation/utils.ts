import * as path from "path";
import * as resolve from "resolve";
import { SourceNode } from "source-map";
import * as ts from "typescript";
// TODO: Don't depend on CLI?
import * as cliDiagnostics from "../cli/diagnostics";
import * as lua from "../LuaAST";
import * as diagnosticFactories from "./diagnostics";

export interface EmitHost {
    directoryExists(path: string): boolean;
    fileExists(path: string): boolean;
    getCurrentDirectory(): string;
    readFile(path: string): string | undefined;
    writeFile: ts.WriteFileCallback;
}

interface BaseFile {
    code: string;
    sourceMap?: string;
    sourceFiles?: ts.SourceFile[];
}

export interface ProcessedFile extends BaseFile {
    fileName: string;
    luaAst?: lua.File;
    /** @internal */
    sourceMapNode?: SourceNode;
}

export interface EmitFile extends BaseFile {
    outputPath: string;
}

export const getConfigDirectory = (options: ts.CompilerOptions) =>
    options.configFilePath ? path.dirname(options.configFilePath) : process.cwd();

export function resolvePlugin(
    kind: string,
    optionName: string,
    basedir: string,
    query: string,
    importName = "default"
): { error?: ts.Diagnostic; result?: unknown } {
    if (typeof query !== "string") {
        return { error: cliDiagnostics.compilerOptionRequiresAValueOfType(optionName, "string") };
    }

    let resolved: string;
    try {
        resolved = resolve.sync(query, { basedir, extensions: [".js", ".ts", ".tsx"] });
    } catch (err) {
        if (err.code !== "MODULE_NOT_FOUND") throw err;
        return { error: diagnosticFactories.couldNotResolveFrom(kind, query, basedir) };
    }

    const hasNoRequireHook = require.extensions[".ts"] === undefined;
    if (hasNoRequireHook && (resolved.endsWith(".ts") || resolved.endsWith(".tsx"))) {
        try {
            const tsNodePath = resolve.sync("ts-node", { basedir });
            const tsNode: typeof import("ts-node") = require(tsNodePath);
            tsNode.register({ transpileOnly: true });
        } catch (err) {
            if (err.code !== "MODULE_NOT_FOUND") throw err;
            return { error: diagnosticFactories.toLoadItShouldBeTranspiled(kind, query) };
        }
    }

    const commonjsModule = require(resolved);
    const factoryModule = commonjsModule.__esModule ? commonjsModule : { default: commonjsModule };
    const result = factoryModule[importName];
    if (result === undefined) {
        return { error: diagnosticFactories.shouldHaveAExport(kind, query, importName) };
    }

    return { result };
}
