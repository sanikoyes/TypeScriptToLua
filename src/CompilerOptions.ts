import * as ts from "typescript";
import * as diagnosticFactories from "./transpilation/diagnostics";

type OmitIndexSignature<T> = {
    [K in keyof T as string extends K ? never : number extends K ? never : K]: T[K];
};

export interface TransformerImport {
    transform: string;
    import?: string;
    after?: boolean;
    afterDeclarations?: boolean;
    type?: "program" | "config" | "checker" | "raw" | "compilerOptions";
    [option: string]: any;
}

export interface LuaPluginImport {
    name: string;
    import?: string;
    [option: string]: any;
}

export type CompilerOptions = OmitIndexSignature<ts.CompilerOptions> & {
    buildMode?: BuildMode;
    noImplicitSelf?: boolean;
    noHeader?: boolean;
    luaBundle?: string;
    luaBundleEntry?: string;
    luaTarget?: LuaTarget;
    luaLibImport?: LuaLibImportKind;
    sourceMapTraceback?: boolean;
    luaPlugins?: LuaPluginImport[];
    plugins?: Array<ts.PluginImport | TransformerImport>;
    tstlVerbose?: boolean;
    [option: string]: any;
};

export enum LuaLibImportKind {
    None = "none",
    Always = "always",
    Inline = "inline",
    Require = "require",
}

export enum LuaTarget {
    Universal = "universal",
    Lua51 = "5.1",
    Lua52 = "5.2",
    Lua53 = "5.3",
    Lua54 = "5.4",
    LuaJIT = "JIT",
}

export enum BuildMode {
    Default = "default",
    Library = "library",
}

export const isBundleEnabled = (options: CompilerOptions) =>
    options.luaBundle !== undefined && options.luaBundleEntry !== undefined;

export function validateOptions(options: CompilerOptions): ts.Diagnostic[] {
    const diagnostics: ts.Diagnostic[] = [];

    if (options.luaBundle && !options.luaBundleEntry) {
        diagnostics.push(diagnosticFactories.luaBundleEntryIsRequired());
    }

    if (options.luaBundle && options.luaLibImport === LuaLibImportKind.Inline) {
        diagnostics.push(diagnosticFactories.usingLuaBundleWithInlineMightGenerateDuplicateCode());
    }

    if (options.luaBundle && options.buildMode === BuildMode.Library) {
        diagnostics.push(diagnosticFactories.cannotBundleLibrary());
    }

    return diagnostics;
}
