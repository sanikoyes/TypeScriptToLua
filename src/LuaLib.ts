import * as path from "path";
import { EmitHost } from "./transpilation";

export enum LuaLibFeature {
    ArrayConcat = "ArrayConcat",
    ArrayEntries = "ArrayEntries",
    ArrayEvery = "ArrayEvery",
    ArrayFilter = "ArrayFilter",
    ArrayForEach = "ArrayForEach",
    ArrayFind = "ArrayFind",
    ArrayFindIndex = "ArrayFindIndex",
    ArrayIncludes = "ArrayIncludes",
    ArrayIndexOf = "ArrayIndexOf",
    ArrayIsArray = "ArrayIsArray",
    ArrayJoin = "ArrayJoin",
    ArrayMap = "ArrayMap",
    ArrayPush = "ArrayPush",
    ArrayReduce = "ArrayReduce",
    ArrayReduceRight = "ArrayReduceRight",
    ArrayReverse = "ArrayReverse",
    ArrayShift = "ArrayShift",
    ArrayUnshift = "ArrayUnshift",
    ArraySort = "ArraySort",
    ArraySlice = "ArraySlice",
    ArraySome = "ArraySome",
    ArraySplice = "ArraySplice",
    ArrayToObject = "ArrayToObject",
    ArrayFlat = "ArrayFlat",
    ArrayFlatMap = "ArrayFlatMap",
    ArraySetLength = "ArraySetLength",
    Class = "Class",
    ClassExtends = "ClassExtends",
    CloneDescriptor = "CloneDescriptor",
    Decorate = "Decorate",
    DecorateParam = "DecorateParam",
    Delete = "Delete",
    DelegatedYield = "DelegatedYield",
    Error = "Error",
    FunctionBind = "FunctionBind",
    Generator = "Generator",
    InstanceOf = "InstanceOf",
    InstanceOfObject = "InstanceOfObject",
    Iterator = "Iterator",
    Map = "Map",
    MathAtan2 = "MathAtan2",
    New = "New",
    Number = "Number",
    NumberIsFinite = "NumberIsFinite",
    NumberIsNaN = "NumberIsNaN",
    NumberToString = "NumberToString",
    ObjectAssign = "ObjectAssign",
    ObjectDefineProperty = "ObjectDefineProperty",
    ObjectEntries = "ObjectEntries",
    ObjectFromEntries = "ObjectFromEntries",
    ObjectGetOwnPropertyDescriptor = "ObjectGetOwnPropertyDescriptor",
    ObjectGetOwnPropertyDescriptors = "ObjectGetOwnPropertyDescriptors",
    ObjectKeys = "ObjectKeys",
    ObjectRest = "ObjectRest",
    ObjectValues = "ObjectValues",
    ParseFloat = "ParseFloat",
    ParseInt = "ParseInt",
    Set = "Set",
    SetDescriptor = "SetDescriptor",
    WeakMap = "WeakMap",
    WeakSet = "WeakSet",
    SourceMapTraceBack = "SourceMapTraceBack",
    Spread = "Spread",
    StringAccess = "StringAccess",
    StringCharAt = "StringCharAt",
    StringCharCodeAt = "StringCharCodeAt",
    StringConcat = "StringConcat",
    StringEndsWith = "StringEndsWith",
    StringIncludes = "StringIncludes",
    StringPadEnd = "StringPadEnd",
    StringPadStart = "StringPadStart",
    StringReplace = "StringReplace",
    StringSlice = "StringSlice",
    StringSplit = "StringSplit",
    StringStartsWith = "StringStartsWith",
    StringSubstr = "StringSubstr",
    StringSubstring = "StringSubstring",
    StringTrim = "StringTrim",
    StringTrimEnd = "StringTrimEnd",
    StringTrimStart = "StringTrimStart",
    Symbol = "Symbol",
    SymbolRegistry = "SymbolRegistry",
    TypeOf = "TypeOf",
    Unpack = "Unpack",
}

/* eslint-disable @typescript-eslint/naming-convention */
const luaLibDependencies: Partial<Record<LuaLibFeature, LuaLibFeature[]>> = {
    ArrayConcat: [LuaLibFeature.ArrayIsArray],
    ArrayFlat: [LuaLibFeature.ArrayConcat, LuaLibFeature.ArrayIsArray],
    ArrayFlatMap: [LuaLibFeature.ArrayConcat, LuaLibFeature.ArrayIsArray],
    Decorate: [LuaLibFeature.CloneDescriptor],
    Delete: [LuaLibFeature.ObjectGetOwnPropertyDescriptors],
    Error: [LuaLibFeature.New, LuaLibFeature.Class],
    FunctionBind: [LuaLibFeature.Unpack],
    Generator: [LuaLibFeature.Symbol],
    InstanceOf: [LuaLibFeature.Symbol],
    Iterator: [LuaLibFeature.Symbol],
    ObjectDefineProperty: [LuaLibFeature.CloneDescriptor, LuaLibFeature.SetDescriptor],
    ObjectFromEntries: [LuaLibFeature.Iterator, LuaLibFeature.Symbol],
    Map: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    Set: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    WeakMap: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    WeakSet: [LuaLibFeature.InstanceOf, LuaLibFeature.Iterator, LuaLibFeature.Symbol, LuaLibFeature.Class],
    Spread: [LuaLibFeature.Iterator, LuaLibFeature.Unpack],
    StringSplit: [LuaLibFeature.StringSubstring, LuaLibFeature.StringAccess],
    SymbolRegistry: [LuaLibFeature.Symbol],
};
/* eslint-enable @typescript-eslint/naming-convention */

export function loadLuaLibFeatures(features: Iterable<LuaLibFeature>, emitHost: EmitHost): string {
    let result = "";

    const loadedFeatures = new Set<LuaLibFeature>();

    function load(feature: LuaLibFeature): void {
        if (loadedFeatures.has(feature)) return;
        loadedFeatures.add(feature);

        const dependencies = luaLibDependencies[feature];
        if (dependencies) {
            dependencies.forEach(load);
        }

        const featurePath = path.resolve(__dirname, `../dist/lualib/${feature}.lua`);
        const luaLibFeature = emitHost.readFile(featurePath);
        if (luaLibFeature !== undefined) {
            result += luaLibFeature + "\n";
        } else {
            throw new Error(`Could not load lualib feature from '${featurePath}'`);
        }
    }

    for (const feature of features) {
        load(feature);
    }

    return `local _ENV = setmetatable({}, { __index = _G })
pcall(function() setfenv(1, _ENV) end)

${result}

return _ENV
`;
}

let luaLibBundleContent: string;
export function getLuaLibBundle(emitHost: EmitHost): string {
    if (luaLibBundleContent === undefined) {
        const lualibPath = path.resolve(__dirname, "../dist/lualib/lualib_bundle.lua");
        const result = emitHost.readFile(lualibPath);
        if (result !== undefined) {
            luaLibBundleContent = result;
        } else {
            throw new Error(`Could not load lualib bundle from '${lualibPath}'`);
        }
    }

    return luaLibBundleContent;
}
