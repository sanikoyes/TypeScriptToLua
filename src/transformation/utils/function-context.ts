import * as ts from "typescript";
import { CompilerOptions } from "../../CompilerOptions";
import { TransformationContext } from "../context";
import { AnnotationKind, getFileAnnotations, getNodeAnnotations } from "./annotations";
import { findFirstNodeAbove, getAllCallSignatures, inferAssignedType } from "./typescript";

export enum ContextType {
    None,
    Void,
    NonVoid,
    Mixed,
}

function hasNoSelfAncestor(declaration: ts.Declaration): boolean {
    const scopeDeclaration = findFirstNodeAbove(
        declaration,
        (node): node is ts.SourceFile | ts.ModuleDeclaration => ts.isSourceFile(node) || ts.isModuleDeclaration(node)
    );

    if (!scopeDeclaration) {
        return false;
    } else if (ts.isSourceFile(scopeDeclaration)) {
        return getFileAnnotations(scopeDeclaration).has(AnnotationKind.NoSelfInFile);
    } else if (getNodeAnnotations(scopeDeclaration).has(AnnotationKind.NoSelf)) {
        return true;
    } else {
        return hasNoSelfAncestor(scopeDeclaration);
    }
}

function getExplicitThisParameter(signatureDeclaration: ts.SignatureDeclaration): ts.ParameterDeclaration | undefined {
    return signatureDeclaration.parameters.find(
        param => ts.isIdentifier(param.name) && param.name.originalKeywordKind === ts.SyntaxKind.ThisKeyword
    );
}

export function getDeclarationContextType(
    { program }: TransformationContext,
    signatureDeclaration: ts.SignatureDeclaration
): ContextType {
    const thisParameter = getExplicitThisParameter(signatureDeclaration);
    if (thisParameter) {
        // Explicit 'this'
        return thisParameter.type && thisParameter.type.kind === ts.SyntaxKind.VoidKeyword
            ? ContextType.Void
            : ContextType.NonVoid;
    }

    // noSelf declaration on function signature
    if (getNodeAnnotations(signatureDeclaration).has(AnnotationKind.NoSelf)) {
        return ContextType.Void;
    }

    if (
        ts.isMethodSignature(signatureDeclaration) ||
        ts.isMethodDeclaration(signatureDeclaration) ||
        ts.isConstructSignatureDeclaration(signatureDeclaration) ||
        ts.isConstructorDeclaration(signatureDeclaration) ||
        (signatureDeclaration.parent && ts.isPropertyDeclaration(signatureDeclaration.parent)) ||
        (signatureDeclaration.parent && ts.isPropertySignature(signatureDeclaration.parent))
    ) {
        // Class/interface methods only respect @noSelf on their parent
        const scopeDeclaration = findFirstNodeAbove(
            signatureDeclaration,
            (n): n is ts.ClassLikeDeclaration | ts.InterfaceDeclaration =>
                ts.isClassDeclaration(n) || ts.isClassExpression(n) || ts.isInterfaceDeclaration(n)
        );

        if (scopeDeclaration === undefined) {
            return ContextType.NonVoid;
        }

        if (getNodeAnnotations(scopeDeclaration).has(AnnotationKind.NoSelf)) {
            return ContextType.Void;
        }

        return ContextType.NonVoid;
    }

    // When using --noImplicitSelf and the signature is defined in a file targeted by the program apply the @noSelf rule.
    const options = program.getCompilerOptions() as CompilerOptions;
    if (options.noImplicitSelf && program.getRootFileNames().includes(signatureDeclaration.getSourceFile().fileName)) {
        return ContextType.Void;
    }

    // Walk up to find @noSelf or @noSelfInFile
    if (hasNoSelfAncestor(signatureDeclaration)) {
        return ContextType.Void;
    }

    return ContextType.NonVoid;
}

function reduceContextTypes(contexts: ContextType[]): ContextType {
    const reducer = (a: ContextType, b: ContextType) => {
        if (a === ContextType.None) {
            return b;
        } else if (b === ContextType.None) {
            return a;
        } else if (a !== b) {
            return ContextType.Mixed;
        } else {
            return a;
        }
    };

    return contexts.reduce(reducer, ContextType.None);
}

function getSignatureDeclarations(
    context: TransformationContext,
    signatures: readonly ts.Signature[]
): ts.SignatureDeclaration[] {
    return signatures.flatMap(signature => {
        const signatureDeclaration = signature.getDeclaration();
        let inferredType: ts.Type | undefined;
        if (
            ts.isMethodDeclaration(signatureDeclaration) &&
            ts.isObjectLiteralExpression(signatureDeclaration.parent) &&
            !getExplicitThisParameter(signatureDeclaration)
        ) {
            inferredType = context.checker.getContextualTypeForObjectLiteralElement(signatureDeclaration);
        } else if (
            (ts.isFunctionExpression(signatureDeclaration) || ts.isArrowFunction(signatureDeclaration)) &&
            !getExplicitThisParameter(signatureDeclaration)
        ) {
            // Infer type of function expressions/arrow functions
            inferredType = inferAssignedType(context, signatureDeclaration);
        }

        if (inferredType) {
            const inferredSignatures = getAllCallSignatures(inferredType);
            if (inferredSignatures.length > 0) {
                return inferredSignatures.map(s => s.getDeclaration());
            }
        }

        return signatureDeclaration;
    });
}

export function getFunctionContextType(context: TransformationContext, type: ts.Type): ContextType {
    if (type.isTypeParameter()) {
        type = type.getConstraint() ?? type;
    }

    if (type.isUnion()) {
        return reduceContextTypes(type.types.map(t => getFunctionContextType(context, t)));
    }

    const signatures = context.checker.getSignaturesOfType(type, ts.SignatureKind.Call);
    if (signatures.length === 0) {
        return ContextType.None;
    }

    const signatureDeclarations = getSignatureDeclarations(context, signatures);
    return reduceContextTypes(signatureDeclarations.map(s => getDeclarationContextType(context, s)));
}
