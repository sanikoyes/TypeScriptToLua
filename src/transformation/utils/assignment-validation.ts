import * as ts from "typescript";
import { getOrUpdate } from "../../utils";
import { TransformationContext } from "../context";
import {
    unsupportedNoSelfFunctionConversion,
    unsupportedOverloadAssignment,
    unsupportedSelfFunctionConversion,
} from "./diagnostics";
import { ContextType, getFunctionContextType } from "./function-context";

// TODO: Clear if types are reused between compilations
const typeValidationCache = new WeakMap<ts.Type, Set<ts.Type>>();

export function validateAssignment(
    context: TransformationContext,
    node: ts.Node,
    fromType: ts.Type,
    toType: ts.Type,
    toName?: string
): void {
    if (toType === fromType) {
        return;
    }

    if ((toType.flags & ts.TypeFlags.Any) !== 0) {
        // Assigning to un-typed variable
        return;
    }

    // Use cache to avoid repeating check for same types (protects against infinite loop in recursive types)
    const fromTypeCache = getOrUpdate(typeValidationCache, fromType, () => new Set());
    if (fromTypeCache.has(toType)) return;
    fromTypeCache.add(toType);

    validateFunctionAssignment(context, node, fromType, toType, toName);

    const fromTypeNode = context.checker.typeToTypeNode(fromType, undefined, undefined);
    const toTypeNode = context.checker.typeToTypeNode(toType, undefined, undefined);
    if (!fromTypeNode || !toTypeNode) {
        return;
    }

    if (
        (ts.isArrayTypeNode(toTypeNode) || ts.isTupleTypeNode(toTypeNode)) &&
        (ts.isArrayTypeNode(fromTypeNode) || ts.isTupleTypeNode(fromTypeNode))
    ) {
        // Recurse into arrays/tuples
        const fromTypeArguments = (fromType as ts.TypeReference).typeArguments;
        const toTypeArguments = (toType as ts.TypeReference).typeArguments;

        if (fromTypeArguments === undefined || toTypeArguments === undefined) {
            return;
        }

        const count = Math.min(fromTypeArguments.length, toTypeArguments.length);
        for (let i = 0; i < count; ++i) {
            validateAssignment(context, node, fromTypeArguments[i], toTypeArguments[i], toName);
        }
    }

    if (
        (toType.flags & ts.TypeFlags.Object) !== 0 &&
        (ts.isTypeLiteralNode(toTypeNode) ||
            ((toType as ts.ObjectType).objectFlags & ts.ObjectFlags.ClassOrInterface) !== 0) &&
        toType.symbol &&
        toType.symbol.members &&
        fromType.symbol &&
        fromType.symbol.members
    ) {
        // Recurse into interfaces
        toType.symbol.members.forEach((toMember, escapedMemberName) => {
            if (fromType.symbol.members) {
                const fromMember = fromType.symbol.members.get(escapedMemberName);
                if (fromMember) {
                    const toMemberType = context.checker.getTypeOfSymbolAtLocation(toMember, node);
                    const fromMemberType = context.checker.getTypeOfSymbolAtLocation(fromMember, node);
                    const memberName = ts.unescapeLeadingUnderscores(escapedMemberName);
                    validateAssignment(
                        context,
                        node,
                        fromMemberType,
                        toMemberType,
                        toName ? `${toName}.${memberName}` : memberName
                    );
                }
            }
        });
    }
}

function validateFunctionAssignment(
    context: TransformationContext,
    node: ts.Node,
    fromType: ts.Type,
    toType: ts.Type,
    toName?: string
): void {
    const fromContext = getFunctionContextType(context, fromType);
    const toContext = getFunctionContextType(context, toType);

    if (fromContext === ContextType.Mixed || toContext === ContextType.Mixed) {
        context.diagnostics.push(unsupportedOverloadAssignment(node, toName));
    } else if (fromContext !== toContext && fromContext !== ContextType.None && toContext !== ContextType.None) {
        context.diagnostics.push(
            toContext === ContextType.Void
                ? unsupportedNoSelfFunctionConversion(node, toName)
                : unsupportedSelfFunctionConversion(node, toName)
        );
    }
}
