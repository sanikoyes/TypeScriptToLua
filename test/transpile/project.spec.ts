import * as path from "path";
import { normalizeSlashes } from "../../src/utils";
import * as util from "../util";

test("should transpile", () => {
    const projectDir = path.join(__dirname, "project");
    const { transpiledFiles } = util
        .testProject(path.join(projectDir, "tsconfig.json"))
        .setMainFileName(path.join(projectDir, "index.ts"))
        .expectToHaveNoDiagnostics()
        .getLuaResult();

    expect(
        transpiledFiles.map(f => ({ filePath: path.relative(projectDir, f.outPath), lua: f.lua }))
    ).toMatchSnapshot();
});

test("should give verbose output", () => {
    // Capture console logs
    const consoleLogs: string[] = [];
    const originalLog = console.log;
    console.log = (...args: any[]) => {
        consoleLogs.push(args.map(a => a.toString().replace(normalizeSlashes(process.cwd()), "<cwd>")).join(","));
    };

    const projectDir = path.join(__dirname, "project");
    util.testProject(path.join(projectDir, "tsconfig.json"))
        .setMainFileName(path.join(projectDir, "index.ts"))
        .setOptions({ tstlVerbose: true })
        .expectToHaveNoDiagnostics();

    console.log = originalLog;

    expect(consoleLogs).toMatchSnapshot();
});
