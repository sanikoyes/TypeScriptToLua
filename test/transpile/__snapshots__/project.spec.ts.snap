// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should give verbose output 1`] = `
Array [
  "Parsing project settings",
  "Successfully loaded 0 plugins",
  "Transforming <cwd>/test/transpile/project/otherFile.ts",
  "Printing <cwd>/test/transpile/project/otherFile.ts",
  "Transforming <cwd>/test/transpile/project/index.ts",
  "Printing <cwd>/test/transpile/project/index.ts",
  "Constructing emit plan",
  "Resolving dependencies for <cwd>/test/transpile/project/otherFile.ts",
  "Resolving dependencies for <cwd>/test/transpile/project/index.ts",
  "Resolving \\"./otherFile\\" from <cwd>/test/transpile/project",
  "Resolved ./otherFile to <cwd>/test/transpile/project/otherFile.ts",
  "Emitting output",
  "Emitting <cwd>/test/transpile/project/otherFile.lua",
  "Emitting <cwd>/test/transpile/project/index.lua",
  "Emit finished!",
]
`;

exports[`should transpile 1`] = `
Array [
  Object {
    "filePath": "otherFile.lua",
    "lua": "local ____exports = {}
function ____exports.getNumber(self)
    return getAPIValue()
end
return ____exports
",
  },
  Object {
    "filePath": "index.lua",
    "lua": "local ____exports = {}
local ____otherFile = require(\\"otherFile\\")
local getNumber = ____otherFile.getNumber
local myNumber = getNumber(nil)
setAPIValue(myNumber * 5)
return ____exports
",
  },
]
`;
